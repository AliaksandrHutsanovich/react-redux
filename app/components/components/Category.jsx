import React from 'react';
import styled from 'styled-components';
import {Link} from 'react-router-dom';

class Category extends React.Component {
    constructor(props) {
        super(props);
        this.state={title: "", tasks:[], subCategories: [], visible: "", width: 0};
        this.count=1;
    }
    render() {

        const Div4=styled.div`
            display: inline-block;
            margin-left: 10px;
            width: 100px;
            height: 25px;
            overflow: hidden;
        `;

        const Button1=styled.button`
            width: 20px;
            height: 20px;
            background-size: 16px 16px;
        `;

        const DecorLink=styled(Link)`
            float: right;
            font-style: italic;
            text-decoration: none;
            margin-top: 3px;
        `;

        const Div3=styled.div`
            background: linear-gradient(to top left, red, #ff65d8);
            width: ${()=>{
            return this.props.width+"%"}
            };
            margin-left: ${()=>{
            return (95-this.props.width)+"%"}
            };
            height: 30px;
            :hover {
                background: green;
            }
            ${document.location.href.includes("/"+this.props.title)}
                background: linear-gradient(to top left, red, #ff65d8);
            ${!document.location.href.includes("/"+this.props.title)}
                background: olive;
        `;

        const Button2=Button1.extend`
            float: right;
            background-size: 16px 16px;
            margin-top: 4px;
        `;

        const Button3=Button1.extend`
            float: right;
            background-size: 16px 16px;
            margin-top: 4px;
        `;

        const Button4=Button1.extend`
            ${this.props.subCategories.length!=0}
               &{
                  display: none;
               }
            background-size: 16px 16px;
        `;

        return (
            <Div3 className={this.props.title} id={this.props.visible}>
                <Button4 className="visibility" onClick={()=>{
                    if (this.count%2!=0){

                        this.props.makeVisibleSubCategory(this.props.title, "yes");
                        this.count=this.count+1;
                    }
                    if (document.getElementsByClassName(this.props.subCategories[0].title)[0].getAttribute("id")=="yes") {
                        this.props.makeUnvisibleSubCategory(this.props.title, "not");
                        this.count=this.count+1;
                    }
                }}></Button4>
                <Div4>
                    <DecorLink to={`/${this.props.title}`} >{this.props.title}</DecorLink>
                </Div4>
                <Button1 className="redact" onClick={()=>{
                    var titleCategory = prompt("Input the title of Category");
                    console.log(this.props.editCategory);
                    var example=this.props.title;
                    if (titleCategory){
                        this.props.editCategory(titleCategory, example);
                    }
                }}></Button1>
                <Button3 className="add_subCategory" onClick={()=>{
                    var example=this.props.title;
                    var event=1;
                    var titleCategory = prompt("Input the title of Category");
                    if (titleCategory) {
                        this.props.addSubCategory({title: titleCategory, subCategories: [], tasks: [], visible: "not"}, example);
                    }
                }}></Button3>
                <Button2 className="delete" onClick={()=>{
                    var example=this.props.title;
                    this.props.deleteCategory(example);
                }}></Button2>
            </Div3>
        );
    }
}

export default Category
