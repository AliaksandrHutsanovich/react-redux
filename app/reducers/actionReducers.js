import {
    addCategory,
    addSubCategory,
    deleteCategory,
    editCategory,
    editTaskStatus,
    addTask,
    editTask,
    deleteTask,
    insertCategory,
    insertTask,
    editCategoryRevive,
    editTaskRevive
} from '../actions/actions';
import { handleActions } from 'redux-actions';
import immutable from 'immutable';
const Map = immutable.Map,
    List = immutable.List;
import { initialState } from './states/initialState';

export const actionReducers = handleActions({

    [addCategory]: {
        next(state, { payload }) {
            return state
                .update("categories",
                (categories) => categories.push(Map({ title: payload.title, subCategories: List([]), tasks: List([]) })));
        }
    },

    [addSubCategory]: {
        next(state, { payload }) {
            return state
                .updateIn(payload.path,
                subCategories => subCategories.push(Map({ title: payload.title, subCategories: List([]), tasks: List([]) })));
        }
    },

    [deleteCategory]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state.deleteIn(fullPath);
        }
    },

    [editCategory]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                category => {
                    return category.set("title", payload.title);
                });
        }
    },

    [editTaskStatus]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                task => {
                    return task.set("isFinished", payload.value);
                });
        }
    },

    [addTask]: {
        next(state, { payload }) {
            return state
                .updateIn(payload.path,
                tasks => tasks.push(Map({ title: payload.title, Description: "", isFinished: false })));
        }
    },

    [editTask]: {
        next(state, { payload }) {
            if (payload.newPath) {
                if (payload.newPathParam) {
                    state = state
                        .updateIn(payload.newPath,
                        tasks => tasks.insert(payload.newPathParam, Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
                }
                else {
                    state = state
                        .updateIn(payload.newPath,
                        tasks => tasks.push(Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
                }
                if (payload.oldPath) {
                    let fullOldPath = Object.assign([], payload.oldPath);
                    fullOldPath.push(payload.oldPathParam);
                    return state.deleteIn(fullOldPath);
                }
                else {
                    return state;
                }
            }
            else {
                let fullOldPath = Object.assign([], payload.oldPath);
                fullOldPath.push(payload.oldPathParam);
                return state
                    .updateIn(fullOldPath,
                    task => task.set('title', payload.title).set('Description', payload.description).set('isFinished', payload.isFinished));
            }
        }
    },

    [deleteTask]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state.deleteIn(fullPath);
        }
    },

    [insertCategory]: {
        next(state, { payload }) {
            return state.updateIn(payload.path, categories => categories.insert(payload.pathParam, Map({ title: payload.title, subCategories: payload.subCategories, tasks: payload.tasks })));
        }
    },

    [insertTask]: {
        next(state, { payload }) {
            return state.updateIn(payload.path, tasks => tasks.insert(payload.pathParam, Map({ title: payload.title, Description: payload.description, isFinished: payload.isFinished })));
        }
    },

    [editCategoryRevive]: {
        next(state, { payload }) {
            let fullPath = Object.assign([], payload.path);
            fullPath.push(payload.pathParam);
            return state
                .updateIn(fullPath,
                category => {
                    return category.set("title", payload.titlePrimary);
                });
        }
    },

    [editTaskRevive]: {
        next(state, { payload }) {
            if (payload.newPath) {
                let fullNewPath = Object.assign([], payload.newPath);
                fullNewPath.push(payload.newPathParam);

                return state
                    .updateIn(payload.oldPath,
                    tasks => tasks.insert(payload.oldPathParam, Map({ title: payload.titlePrimary, Description: payload.descriptionPrimary, isFinished: payload.isFinishedValuePrimary })))
                    .deleteIn(fullNewPath);
            }
            else {
                let fullOldPath = Object.assign([], payload.oldPath);
                fullOldPath.push(payload.oldPathParam);
                return state.updateIn(fullOldPath, task => task.set('title', payload.titlePrimary).set('Description', payload.descriptionPrimary).set('isFinished', payload.isFinishedValuePrimary));
            }
        }
    }

}, Map().merge(initialState));