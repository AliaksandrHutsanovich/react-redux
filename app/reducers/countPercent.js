import { countPercentage, incrementInDone, decrementInDone, incrementInAll, decrementInAll } from '../actions/actions';
import { handleActions } from 'redux-actions';
import immutable from 'immutable';
const Map = immutable.Map;

export const countPer—Åent = handleActions({

    [countPercentage]: {
        next(state, { payload }) {
            return state.set('done', payload.done).set('all', payload.all);
        }
    },

    [incrementInDone]: {
        next(state, { payload }) {
            let done = state.get('done') + 1;
            return state.set('done', done);
        }
    },

    [decrementInDone]: {
        next(state, { payload }) {
            let done = state.get('done') - 1;
            return state.set('done', done);
        }
    },

    [incrementInAll]: {
        next(state, { payload }) {
            let all = state.get('all') + 1;
            return state.set('all', all);
        }
    },

    [decrementInAll]: {
        next(state, { payload }) {
            let all = state.get('all') - 1;
            return state.set('all', all);
        }
    }

}, Map({ done: 0, all: 0 }));

