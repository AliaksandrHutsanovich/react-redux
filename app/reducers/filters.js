import { Map } from 'immutable';
import { handleActions } from 'redux-actions';
import {
  showDoneTasks,
  searchTasks,
} from '../actions';

const filters = handleActions({

  [showDoneTasks]: {
    next(state, { payload }) {
      return state.set('showDone', payload);
    },
  },

  [searchTasks]: {
    next(state, { payload }) {
      return state.set('searchKey', payload);
    },
  },

}, Map({ showDone: true, searchKey: '' }));

export default filters;
