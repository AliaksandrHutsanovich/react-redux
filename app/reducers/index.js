export { default } from './reducer';
export { initialState } from './states/initialState';
