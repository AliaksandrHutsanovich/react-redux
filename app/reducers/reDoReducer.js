import { Map, List } from 'immutable';
import { handleActions } from 'redux-actions';
import { addToReDo, deleteFromReDo, clearReDo } from '../actions';

const reDoReducer = handleActions({

  [addToReDo]: {
    next(state, { payload }) {
      return state
        .update('redoOperations', (operations) => operations.push(Map(payload)));
    },
  },

  [deleteFromReDo]: {
    next(state) {
      return state.update('redoOperations', (operations) => operations.delete(state.get('redoOperations').toArray().length - 1));
    },
  },

  [clearReDo]: {
    next(state) {
      return state.set('redoOperations', List([]));
    },
  },

}, Map({ redoOperations: List([]) }));

export default reDoReducer;
