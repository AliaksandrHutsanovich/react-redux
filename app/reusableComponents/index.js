export { default as AddForm } from './addForm';
export { default as ChangingDataDialog } from './changingDataModal';
