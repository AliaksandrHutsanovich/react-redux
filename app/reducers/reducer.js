import { combineReducers } from 'redux';
import {actionReducers} from './actionReducers'; 
import {filters} from  './filters';
import { countPerсent } from './countPercent';
import { reDoReducer } from './reDoReducer';
import { unDoReducer } from './unDoReducer';

export default combineReducers({
    actionReducers,
    filters,
    countPerсent,
    reDoReducer,
    unDoReducer
});








