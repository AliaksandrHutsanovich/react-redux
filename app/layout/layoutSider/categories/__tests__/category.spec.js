import React from 'react';
import { shallow } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Map } from 'immutable';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Icon } from 'antd';
import { initialState as store } from '../../../../reducers';
import Category from '../category';
import { ChangingDataDialog } from '../../../../reusableComponents';

const mockStore = configureStore();
const initialState = Map().merge(store);

describe('Unit test of Category component', () => {
  const dataStore = mockStore({ actionReducers: initialState });
  const Component = shallow(
    <Provider store={dataStore}>
      <BrowserRouter>
        <Category
          path="category-0"
          title="My category"
        />
      </BrowserRouter>
    </Provider>,
  );

  const fragment = Component.find(Category).shallow();
  it('Full render test', () => {
    expect(fragment).toMatchSnapshot();
  });
  it('Delete modal should be openable', () => {
    fragment.find(Icon).at(2).simulate('click');
    expect(fragment.find(ChangingDataDialog).at(2).prop('visible')).toBeTruthy();

    fragment.find(ChangingDataDialog).at(2).prop('onOk')();
    expect(fragment.find(ChangingDataDialog).at(2).prop('visible')).toBeFalsy();

    fragment.find(Icon).at(2).simulate('click');
    fragment.find(ChangingDataDialog).at(2).prop('onCancel')();
    expect(fragment.find(ChangingDataDialog).at(2).prop('visible')).toBeFalsy();
  });
  it('Edit modal should be openable and closable', () => {
    fragment.find(Icon).at(0).simulate('click');
    expect(fragment.find(ChangingDataDialog).at(1).prop('visible')).toBeTruthy();

    fragment.find(ChangingDataDialog).at(1).prop('onOk')();
    expect(fragment.find(ChangingDataDialog).at(1).prop('visible')).toBeFalsy();

    fragment.find(Icon).at(0).simulate('click');
    fragment.find(ChangingDataDialog).at(1).prop('onCancel')();
    expect(fragment.find(ChangingDataDialog).at(1).prop('visible')).toBeFalsy();
  });
  it('Add modal should be openable and closable', () => {
    fragment.find(Icon).at(1).simulate('click');
    expect(fragment.find(ChangingDataDialog).at(0).prop('visible')).toBeTruthy();

    fragment.find(ChangingDataDialog).at(0).prop('onOk')();
    expect(fragment.find(ChangingDataDialog).at(0).prop('visible')).toBeFalsy();

    fragment.find(Icon).at(1).simulate('click');
    fragment.find(ChangingDataDialog).at(0).prop('onCancel')();
    expect(fragment.find(ChangingDataDialog).at(0).prop('visible')).toBeFalsy();
  });
});
