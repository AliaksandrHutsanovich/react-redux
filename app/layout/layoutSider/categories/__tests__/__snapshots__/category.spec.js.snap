// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test of Category component Full render test 1`] = `
<div>
  <span>
    <Link
      to="/category-0"
    >
      My category
    </Link>
  </span>
  <Icon
    className="itemButton"
    onClick={[Function]}
    type="edit"
  />
  <Icon
    className="itemButton"
    onClick={[Function]}
    type="plus"
  />
  <Icon
    className="itemButton"
    onClick={[Function]}
    type="delete"
  />
  <Connect(Component)
    formId="addNewSubCategory-My category"
    onCancel={[Function]}
    onOk={[Function]}
    operationTitle="Add new subcategory"
    path="category-0"
    visible={false}
  />
  <Connect(Component)
    formId="editCategory-My category"
    onCancel={[Function]}
    onOk={[Function]}
    operationTitle="Edit category"
    path="category-0"
    title="My category"
    visible={false}
  />
  <Connect(Component)
    onCancel={[Function]}
    onOk={[Function]}
    operationTitle="Delete category"
    path="category-0"
    title="My category"
    visible={false}
  />
</div>
`;
